File: app/src/main/java/fi/kaupunkifillarit/maps/Maps.java
package fi.kaupunkifillarit.maps;

import android.graphics.Bitmap;

import fi.kaupunkifillarit.model.MapLocation;

public abstract class Maps {
    public interface MapWrapper<MW extends Maps.MarkerWrapper, MOW extends Maps.MarkerOptionsWrapper> {
        void animateToMapLocation(MapLocation mapLocation);

        void setMyLocationEnabled(boolean enabled);

        void setTrafficEnabled(boolean enabled);

        void setPadding(int left, int top, int right, int bottom);

        void setOnMarkerClickListener(OnMarkerClickListener listener);

        void setOnMapLocationChangeListener(OnMapLocationChangeListener listener);

        void setOnMyLocationButtonClickListener(OnMyLocationButtonClickListener listener);

        MOW createMarkerOptions();

        MW addMarker(MOW markerOptions);
    }

    public interface MarkerWrapper<T> {
        T getMarker();

        void setPosition(double latitude, double longitude);
    }

    public interface MarkerOptionsWrapper<T> {
        T getMarkerOptions();

        MarkerOptionsWrapper<T> icon(Bitmap bitmap);

        MarkerOptionsWrapper<T> flat(boolean flat);

        MarkerOptionsWrapper<T> position(double latitude, double longitude);

        MarkerOptionsWrapper<T> anchor(float x, float y);
    }

    public interface OnMarkerClickListener {
        boolean onMarkerClick(MarkerWrapper marker);
    }

    public interface OnMapLocationChangeListener {
        void onMapLocationChange(MapLocation mapLocation);
    }

    public interface OnMyLocationButtonClickListener {
        boolean onMyLocationButtonClick();
    }
}

File: app/src/main/kotlin/fi/kaupunkifillarit/maps/Maps.kt
package fi.kaupunkifillarit.maps

import android.graphics.Bitmap

import fi.kaupunkifillarit.model.MapLocation

abstract class Maps {
    interface MapWrapper<out MW : Maps.MarkerWrapper<*>, MOW : Maps.MarkerOptionsWrapper<*>> {
        var myLocationEnabled: Boolean

        var trafficEnabled: Boolean

        fun animateToMapLocation(mapLocation: MapLocation)

        fun setPadding(left: Int, top: Int, right: Int, bottom: Int)

        fun setOnMarkerClickListener(listener: OnMarkerClickListener)

        fun setOnMapLocationChangeListener(listener: OnMapLocationChangeListener)

        fun setOnMyLocationButtonClickListener(listener: OnMyLocationButtonClickListener)

        fun createMarkerOptions(): MOW

        fun addMarker(markerOptions: MOW): MW
    }

    interface MarkerWrapper<out T> {
        val marker: T

        fun remove()
    }

    interface MarkerOptionsWrapper<T> {
        val markerOptions: T

        fun icon(bitmap: Bitmap): MarkerOptionsWrapper<T>

        fun flat(flat: Boolean): MarkerOptionsWrapper<T>

        fun position(latitude: Double, longitude: Double): MarkerOptionsWrapper<T>

        fun anchor(x: Float, y: Float): MarkerOptionsWrapper<T>
    }

    interface OnMarkerClickListener {
        fun onMarkerClick(marker: MarkerWrapper<*>): Boolean
    }

    interface OnMapLocationChangeListener {
        fun onMapLocationChange(mapLocation: MapLocation)
    }

    interface OnMyLocationButtonClickListener {
        fun onMyLocationButtonClick(): Boolean
    }
}
